services/checks/registry.go

```diff
@@ -0,0 +1,76 @@
+// pmm-managed
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package checks
+
+import (
+	"sync"
+	"time"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/percona/pmm/api/alertmanager/ammodels"
+)
+
+// registry stores alerts and delay information by IDs.
+type registry struct {
+	rw     sync.RWMutex
+	alerts map[string]ammodels.PostableAlert
+	nowF   func() time.Time // for tests
+}
+
+// newRegistry creates a new registry.
+func newRegistry() *registry {
+	return &registry{
+		alerts: make(map[string]ammodels.PostableAlert),
+		nowF:   time.Now,
+	}
+}
+
+// createAlert creates alert from given AlertParams
+func (r *registry) createAlert(labels, annotations map[string]string, alertTTL time.Duration) ammodels.PostableAlert {
+	return ammodels.PostableAlert{
+		Alert: ammodels.Alert{
+			// GeneratorURL: "TODO",
+			Labels: labels,
+		},
+		EndsAt:      strfmt.DateTime(r.nowF().Add(alertTTL)),
+		Annotations: annotations,
+	}
+}
+
+// set clears the previous alerts and sets a new slice of alerts in the registry
+func (r *registry) set(alerts []alertWithID) {
```

Comment:
Is there any possibility to have two or more alerts with the same alert id in this slice?

Reply:
the id is calculated like this `id := alertsPrefix + hashID(target.serviceID+"/"+result.Summary)` so if the service.ID and result.Summary  are the same then yes.

the summary can be same in case the user writes a check that sends lets say 2 alerts with the same summary and if these alerts come from any one target then yes I think they will have the same id

```
script: |
          def check(rows):
              results = []
    
              results.append({
                  "summary": "Fake check",
                  "description": "failed check .....",
                  "severity": "warning",
              })
              results.append({
                  "summary": "Fake check",
                  "description": "failed check .....",
                  "severity": "warning",
              })
    
              return results
```

i dont think anyone would write such checks but I guess it can be possible

Reply:
7fd812edb035b08d0d7518ebced375f1000f021b